From aa073a5f93c17e5a3f6a721af1b190e23169fa50 Mon Sep 17 00:00:00 2001
From: yangfl <yangfl@users.noreply.github.com>
Date: Sun, 14 Jun 2020 14:49:28 +0800
Subject: [PATCH 5/8] Force INI_HANDLER_LINENO

---
 meson.build | 1 +
 1 file changed, 1 insertion(+)

diff --git a/meson.build b/meson.build
index c419c6a..91a0d9a 100644
--- a/meson.build
+++ b/meson.build
@@ -13,6 +13,7 @@ extra_args = []
 
 if distro_install
     pkg = import('pkgconfig')
+    arg_static += ['-DINI_HANDLER_LINENO=1']
 else
     if not get_option('multi-line_entries')
         arg_static += ['-DINI_ALLOW_MULTILINE=0']
-- 
2.47.2

