From 9adeab9afc4a7648817071f3ebfefc09d7ad7996 Mon Sep 17 00:00:00 2001
From: yangfl <yangfl@users.noreply.github.com>
Date: Tue, 9 Feb 2021 19:44:21 +0800
Subject: [PATCH 6/8] Fix cpp type of INI_HANDLER_LINENO

---
 cpp/INIReader.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/cpp/INIReader.cpp b/cpp/INIReader.cpp
index 215e68f..46bf92c 100644
--- a/cpp/INIReader.cpp
+++ b/cpp/INIReader.cpp
@@ -12,6 +12,10 @@
 #include <algorithm>
 #include <cctype>
 #include <cstdlib>
+#ifdef INI_HANDLER_LINENO
+#undef INI_HANDLER_LINENO
+#endif
+#define INI_HANDLER_LINENO 0
 #include "../ini.h"
 #include "INIReader.h"
 
-- 
2.47.2

